api/src/main/java/org/pmiops/workbench/api/OfflineClusterController.java

```diff
@@ -0,0 +1,153 @@
+   * upgrades, cost savings, and minimizing user disruption. To this point, our
+   * goal is to only upgrade idle clusters if possible, as doing so gives us an
+   * assurance that a researcher is not actively using it. We delete clusters in
+   * the following cases:
+   *
+   * 1. It exceeds the max cluster age. Per environment, but O(weeks).
+   * 2. It is idle and exceeds the max idle cluster age. Per environment,
+   *    smaller than (1).
+   *
+   * As an App Engine cron endpoint, the runtime of this method may not exceed
+   * 10 minutes.
+   */
+  @Override
+  public ResponseEntity<CheckClustersResponse> checkClusters() {
+    Instant now = clock.instant();
+    WorkbenchConfig config = configProvider.get();
+    Duration maxAge = Duration.ofDays(config.firecloud.clusterMaxAgeDays);
+    Duration idleMaxAge = Duration.ofDays(config.firecloud.clusterIdleMaxAgeDays);
+
+    ClusterApi clusterApi = clusterApiProvider.get();
+    List<Cluster> clusters;
+    try {
+      clusters = clusterApi.listClusters(null, false);
+    } catch (ApiException e) {
+      throw ExceptionUtils.convertNotebookException(e);
+    }
+
+    int errors = 0;
+    int activeDeletes = 0;
+    int unusedDeletes = 0;
+    for (Cluster c : clusters) {
+      String clusterId = c.getGoogleProject() + "/" + c.getClusterName();
+      try {
+        // Refetch the cluster to ensure freshness as this iteration may take
+        // some time.
+        c = clusterApi.getCluster(c.getGoogleProject(), c.getClusterName());
+      } catch (ApiException e) {
+        log.log(Level.WARNING, String.format("failed to refetch cluster '%s'", clusterId), e);
+        errors++;
+        continue;
+      }
+      if (ClusterStatus.UNKNOWN.equals(c.getStatus()) || c.getStatus() == null) {
+        log.warning(String.format("unknown cluster status for cluster '%s'", clusterId));
+        continue;
+      }
+      if (!ClusterStatus.RUNNING.equals(c.getStatus()) &&
+          !ClusterStatus.STOPPED.equals(c.getStatus())) {
+        // For now, we only handle running or stopped (suspended) clusters.
+        continue;
+      }
+
+      Instant created = Instant.parse(c.getCreatedDate());
+      Duration age = Duration.between(created, now);
+      Instant lastUsed = Instant.parse(c.getDateAccessed());
+      boolean isIdle = Duration.between(lastUsed, now).toHours() > IDLE_AFTER_HOURS;
+      if (age.toMillis() > maxAge.toMillis()) {
+        log.info(String.format(
+            "deleting cluster '%s', exceeded max lifetime @ %s (>%s)",
+            clusterId, formatDuration(age), formatDuration(maxAge)));
+        activeDeletes++;
+      } else if (isIdle && age.toMillis() > idleMaxAge.toMillis()) {
+        log.info(String.format(
+            "deleting cluster '%s', idle with age %s (>%s)",
+            clusterId, formatDuration(age), formatDuration(idleMaxAge)));
+        unusedDeletes++;
+      } else {
+        // Don't delete.
```

Comment:
It might be useful to spit out how many clusters were idle in general, whether or not we deleted them. (Just as an extra piece of data.)

Reply:
Done